### ISSUE TITLE
API: Utility functions to convert binary addresses into FI_ADDR_STR formatted addresses (and vice versa)

### ISSUE THREAD

**Issue by sydidelot** [ofiwg/libfabric#6016]
Hi, 
I'm looking for a way to convert binary addresses (e.g., `(struct fi_info *)->src_addr`  and (`struct fi_info *)->dest_addr`) into a string representation that can be used as an input argument to `fi_getinfo()`. While this conversion operation is straightforward for `FI_SOCKADDR_IN` and `FI_SOCKADDR_IN6`, it gets more complex for the other providers that have their own internal representation (e.g., `FI_ADDR_IB_UD`, `FI_ADDR_EFA` or `FI_SOCKADDR_IB` after PR #5605 gets merged).

`src/common.c` already defines the functions `ofi_straddr()` and `ofi_str_toaddr()` to convert binary addresses into strings and vice versa. Those functions are however not exposed in the public API so far. Would it be possible to expose them?

**Comment by shefty** [comment]
You can pass the binary address into fi_getinfo() by setting fi_info->src_addr/dest_addr in the hints.  It saves the provider from needing to convert the node and service parameters back into the binary form.

**Comment by sydidelot** [comment]
I have a new use-case where the addresses are transferred between the peers that need to establish a connection using a service that only accepts strings. I can do the binary <-> string conversion into the application, but since Libfabric already implements such a facility, I thought it would have been easier to just expose that to the API.

**Comment by shefty** [comment]
It may not be ideal, but I think the existing fi_tostr() can be used for this.  It requires setting fi_info::addr_format and src_addr or dest_addr fields.  The address string can be extracted and should be usable as the fi_getinfo() node parameter.

A better option would be to define a new structure that only specifies the addr_format and addr field, which could be passed into fi_tostr() with a new FI_TYPE_xxx value defined.

**Comment by VoVAllen** [comment]
Hi,

I met similar problems, why `ofi_str_toaddr` is not exposed to user

**Comment by shefty** [comment]
The call is not exposed because it means maintaining it as a stable API going forward.  One problem with exposing the call directly is that it allocates memory, which the caller must free.  That's not a public API that we want to expose.  It's better for the user to own the buffer allocation.

fi_tostr(), or rather a thread safe version of fi_tostr(), would be a better option.

**Comment by shefty** [comment]
I've merged support for fi_tostr_r().  I can add more direct support to that call for converting a binary address into a string.

The reverse call would still not be exposed, though the functionality is available using fi_getinfo() by passing the string in as the node parameter.

The main drawback I see here is that the conversion is handled by the libfabric core.  If there is an external provider using a proprietary address format, there's no mechanism in place to handle that case.

**Comment by shefty** [comment]
See PR #6356

**Comment by VoVAllen** [comment]
Thanks a lot. The problem I met is `FI_ADDR_STR` is not well supported. If I pass a `FI_ADDR_STR` address as node in `fi_getinfo`, it can only return the `shm` provider (when using RDM endpoint), which is the only provider using FI_ADDR_STR. Therefore it seems I need manually change `FI_ADDR_STR` into the supported format for other provider, but there's no function to do so.

**Comment by VoVAllen** [comment]
Otherwise I have to set up another control endpoint using socket to exchange the address information between machines.

**Comment by u-u-h** [comment]
While I'm happy to see `fi_tostr_r()` I'll venture out on a limb and guess the true intent of the request.
While having a string representation of addresses is a good first approximation to get a printable address, it's not sufficient to use these addresses on other nodes to infer how to connect to the respective endpoint: IP addresses would match both for sockets and tcp providers, but without knowing which one the other end is using, libfabic cannot connect to it. Similarly, for IB we need to distinguish `FI_PROTO_RDMA_CM_IB_RC` and `FI_PROTO_RDMA_CM_IB_XRC`, and here even the provider name is the same, so a uri-style address `verbs://.../` does not automatically include this information.

**Comment by shefty** [comment]
Additional information can be converted into a string an appended, as needed.  fi_tostr() can convert the protocol, though the conversion back is non-trivial.  It may be easier to convert that using a string representation of the numeric value (e.g. "2"), rather than using fi_tostr().  The provider name is already a string.

Updating the providers to support FI_ADDR_STR is a separate issue that needs to be addressed.

**Comment by shefty** [comment]
Depending on what the application is doing here, the proposed struct fi_ep_addr can be extended with the additional information needed (protocol, provider name), with that information added onto the string. Or the application can provide struct fi_info as input, with the output restricted to only what's needed to convey the EP address.

**Comment by shefty** [comment]
I spent some time looking at this.  I *don't* believe we need something like fi_fromstr().

As mentioned, a peer needs additional information to go along with the address in order to setup communication.  What's needed is application specific, but _could_ include capability bits, fabric name, provider name, protocol, endpoint type, and other items.  These fields are either strings already or simple integer values.  Converting integers to/from strings is trivial for an application to do.

The only complex data that I see being needed is the endpoint address, which, hey, is the primary request for this issue.  I think the real issue is that providers are not supporting FI_ADDR_STR properly.  With FI_ADDR_STR, calls such as fi_getinfo()/fi_getname() would automatically return the address back as a string.  That same string could be passed to a peer and used in fi_av_insert() for example, or passed in as the node parameter to fi_getinfo().

PR #6356 exposes ofi_tostr_addr() functionality to the app.  However, the exposing of ofi_str_toaddr() is still not available.  *If* we expose ofi_tostr_addr(), it does seem that exposing ofi_str_toaddr() is required.  But I don't know that we need either if IF_ADDR_STR is supported properly.

**Comment by u-u-h** [comment]
I agree that proper `FI_ADDR_STR` support would provide a very usable path forward that could avoid a lot of the issues discussed here.
I also agree that for proper connection setup more is needed, and it's hard to imagine a universal solution outside a full `serialize_endpoint()` implementation, so it's probably best left to the applications for now.
